---
title: "Estimating permutation p-values using MatrixEQTL"
output:
  pdf_document: default
  html_document: default
  word_document: default
---
In our pipeline we first reformat the data per gene and then for each preprocessed gene run step4_MatrixEQTL script which runs multiple bootstraps.

```{r setup, include=FALSE, echo=FALSE, message=F, results='hide'}
require("knitr")
```
First, load the data in TReCASE format. Arguments here are in the same style as original pipeline script. They give information about the chromosome on which gene is located, number of subsamples to be used for estimation (no more then total number of samples recorded in specification file), random seed, window size and which model to be used.

Specification file still will be used, since it is required at earlier steps linking in this pipeline. It is not necessary if you choose a different way to provide the path to the data.

The data in this example is simulated based on the GTEx dataset which allows to avoid distribution of the real data, but provides a dataset that is well represented of the GTEx dataset.

```{r initial_config, include=TRUE, echo=FALSE, message=F, results='hide'}
args = c("9", "1", "5e+05", "short")
chri = as.numeric(args[1])
geni = as.numeric(args[2])
cis_window = as.numeric(args[3])
model = args[4]
nsub = 150

specf = "specifications_Muscle_Skeletal.txt"
getwd()

specs = unlist(read.table(specf, as.is=T))
specs
pref = specs[1]
nsam = specs[2]
queue = specs[3]
days = specs[4]
bmem = as.numeric(specs[5])
seedval = specs[13]
wrk.dir = specs[14]
lib.dir = specs[15]
bas.dir = specs[16]
asSeq.lib = specs[18]
rcmd = specs[19]

c(pref, nsub)

permute = F
set.seed(seedval)
subs = sample(1:nsam)[1:nsub]
# ------------------------------------------------------------
# read in data
# ------------------------------------------------------------
#root.dir = "/pine/scr/z/h/zhabotyn/R01/GTEx/v8"
#gtex.dir = root.dir

outdir0 = sprintf("run_%s_%s_%s_%s", pref, nsub, cis_window, model)
if(!file.exists(outdir0))dir.create(outdir0)

wrk.dir = getwd()
bas.dir = sprintf("%s/%s", wrk.dir, pref)
cnt.dir = sprintf("%s_prepr", pref);cnt.dir;file.exists(cnt.dir)
int.dir = sprintf("%s/%s_%s_%s", bas.dir, pref, nsam, cis_window)

library(asSeq)


  
```

Once initial setup is done we read relevant gene information. Also, get a subset of 150 to make time to fit more feasible.

```{r reading relevant data, include=TRUE, message=F, results='hide'}
genepos_file_name = sprintf("%s/%s/geneInfo_prepr_%s.txt", 
                    bas.dir, cnt.dir, model, echo=FALSE, message=F, results='hide')
geneInfo = read.table(genepos_file_name, as.is=T,header=T)



eChr = rep(chri, nrow(geneInfo))
ePos = as.numeric((geneInfo[,3] + geneInfo[,4])/2)


suffi = sprintf("%s_%s", chri, geni)
#genotypes_%s.dat
genfi = read.table(sprintf("%s/genotypes_%s.dat", int.dir, suffi), as.is=T, header=T)
infi = read.table(sprintf("%s/genotypei_%s.dat", int.dir, suffi), as.is=T, header=T)
#genfi[1:4,1:4]
table(unlist(genfi[,1]))
head(infi)

cnts = read.csv(sprintf("%s/counti_%s_%s.csv", int.dir, model, suffi), as.is=T, header=F)
#cnts[1:4,]
Xmatfil = sprintf("%s/Xmat_%s.csv", int.dir, model)
X = read.csv(Xmatfil, as.is=T, header=F)
nr = nrow(genfi)
genfi = matrix(unlist(genfi), nrow=nr)
nsam = nrow(X)
X = matrix(unlist(X), nrow=nsam)
dim(X)
#X[1:4,]

X = X[subs,]
cnts = cnts[subs,]
genfi = genfi[,subs]

#do an extra step to ensure that there is no 0-variance covariates
converge = 1e-4
vari = apply(X,2,var)

updvar = which(vari<converge)
for(i in updvar){
  if(length(vari[-updvar]>0)>0){
    correct = sqrt(median(vari[-updvar]))/sqrt(vari[i])
  }else{
    correct = 1/sqrt(vari[i])
  }
  xm = mean(X[,i])    
  X[,i] = xm+(X[,i]-xm)*correct
}
vari
apply(X,2,var)

cnts = matrix(as.numeric(unlist(cnts)), nrow=nsub)
#cnts[1:4,]





```



Fit the model 
```{r fit_the_model, include=TRUE, message=F, results='hide'}
mChr = rep(chri, nr)
#geni = 1
#message("geno: ", nrow(geno), " ", ncol(geno), " trecD: ", nrow(trecD), " ", ncol(trecD), " ", nrow(geneInfo))
#geni = 389

#geni = indi
  local.distances = as.numeric((geneInfo[,4] - geneInfo[,3]))/2+cis_window
#  kp = which(SNPInfo[,3]>=(ePos[geni]-local.distances[geni]) & SNPInfo[,3]<(ePos[geni]+local.distances[geni]));length(kp)
  kp = nrow(infi)
  sum(kp)
  
output.tagi     = sprintf("%s/%s",
                               outdir0, geneInfo[geni,1])#rownames(trecD)[geni])
    timj = sprintf("%s_time.txt", output.tagi)
    res.lon = sprintf("%s_eqtl.txt", output.tagi)
      
    time1 = proc.time()
    asSeq:::trecase(Y=cnts[,1,drop=F], Y1=cnts[,2,drop=F], Y2=cnts[,3,drop=F], X=X, Z=t(genfi), output.tag = output.tagi, 
                p.cut=1,  local.distance = local.distances[geni], 
                eChr = eChr[geni], 
                ePos = ePos[geni], 
                mChr = as.numeric(infi[,2]), mPos = as.numeric(infi[,3]),
                maxit = 4000, 
                min.AS.reads = 5, min.AS.sample = 5, min.n.het = 5)

    time2 = proc.time()
  
    write.table(time2[3]-time1[3], timj, row.names=F, col.names=F, quote=F)
    eqtl = read.table(res.lon, header=T, as.is=T)
  
    eqtl[,1] = infi[eqtl[,2],3]; colnames(eqtl)[1] = "Pos"
    eqtl[,2] = infi[eqtl[,2],1]
    eqtl$chr = chri
    write.table(eqtl, res.lon, row.names=F, col.names=T, quote=F, sep="\t")


```


Results are outputed in the following format:
```{r short_output, include=TRUE}
eqtl[1:5,]
```

Time required to fit the model
```{r time_output, include=TRUE}
time2 - time1 

```

